spec/requests/api/v1/allocation_events_controller_cancel_spec.rb

```diff
@@ -0,0 +1,121 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Api::V1::AllocationEventsController do
+  let(:response_json) { JSON.parse(response.body) }
+
+  describe 'POST /allocations/:id/cancel' do
+    let(:schema) { load_yaml_schema('post_allocation_cancel_responses.yaml') }
+
+    let(:supplier) { create(:supplier) }
+    let(:application) { create(:application, owner_id: supplier.id) }
+    let(:access_token) { create(:access_token, application: application).token }
+    let(:headers) { { 'CONTENT_TYPE': content_type, 'Authorization': "Bearer #{access_token}" } }
+    let(:content_type) { ApiController::CONTENT_TYPE }
+
+    let(:allocation) { create(:allocation, :with_moves) }
+    let(:move) { allocation.moves.first }
+    let(:allocation_id) { allocation.id }
+    let(:cancellation_reason) { 'made_in_error' }
+    let(:cancel_params) do
+      {
+        data: {
+          type: 'cancel',
+          attributes: {
+            cancellation_reason: cancellation_reason,
+            cancellation_reason_comment: 'FUBAR',
+          },
+        },
+      }
+    end
+
+    before do
+      allow(Notifier).to receive(:prepare_notifications)
+      post "/api/v1/allocations/#{allocation_id}/cancel", params: cancel_params, headers: headers, as: :json
+    end
+
+    context 'when successful' do
+      it_behaves_like 'an endpoint that responds with success 204'
+
+      it 'updates the allocation cancellation reason' do
+        expect(allocation.reload.cancellation_reason).to eq(cancellation_reason)
+      end
+
+      it 'updates the allocation cancellation reason comment' do
+        expect(allocation.reload.cancellation_reason_comment).to eq('FUBAR')
+      end
+
+      it 'updates the allocation status' do
+        expect(allocation.reload.status).to eql('cancelled')
+      end
+
+      it 'updates the allocation moves_count' do
+        expect(allocation.reload.moves_count).to eq(0)
+      end
+
+      it 'updates the allocation moves status' do
+        expect(allocation.reload.moves.pluck(:status).uniq).to contain_exactly('cancelled')
+      end
+
+      describe 'webhook and email notifications' do
+        it 'calls the move notifier when updating the status' do
+          expect(Notifier).to have_received(:prepare_notifications).with(topic: move, action_name: 'update_status')
+        end
+      end
+    end
+
+    context 'without specifying cancellation reason comment' do
+      let(:cancel_params) do
+        {
+          data: {
+            type: 'cancel',
+            attributes: {
+              cancellation_reason: cancellation_reason,
+            },
+          },
+        }
+      end
+
+      it 'updates the allocation cancellation reason comment' do
```

Comment:
I'm wondering if we should instead of falling back to the default return a validation error, as initially the default was set until we implemented this 🤔

Reply:
I don't think an error would be appropriate, as the cancellation reason comment is optional. Perhaps instead the reason comment param could be omitted and set as `nil` in both the allocation and move models.

Reply:
I've amended to allow a blank cancellation reason comment to be passed in the request and then set this against both the allocation and move.